Add Recommended Configuration to Task Types

daniel-wer opened the pull request:
This PR allows to add a recommended configuration to task types. The recommended config is shown to users when they start to trace a task and their configuration differs from the recommended one. Users can then accept or decline this config which automatically applies it (or not).

The message will not trigger twice if the "Finish and get next task" functionality is used and the user changes settings in the mean time. It will also not trigger if a tracing is continued. However, if the user returns to the dashboard and starts the next task from there, the message will trigger again, although the task type did not change.
The requirement (from discuss) was:

> These settings would then be set whenever a hiwi get a new task type.
again, only once when the task type changes.

but I'm not sure how to achieve that. We would need to persist the last task type for that.
With the current implementation, I'm afraid users might get annoyed with the configuration recommendations. What do you think?

### URL of deployed dev instance (used for testing):
- https://recommendedsettings.webknossos.xyz

### Steps to test:
- Create a new task type. Select to "Add Recommended User Settings" and play around with the values, json in the textarea. You should not be able to create the task type with an invalid configuration.
- Once you created a task type with a recommended config, click "Edit" in the task type list. The recommended config field should be checked and expanded.
- During task type creation, if the recommended config contains errors and you uncheck that you want to add a recommended config, you should still be able to create the task type.

- Create a task with a task type with/without a recommended config and get this task as a user.
- Open the task to trace. You should see a message informing you about the recommended settings that can be applied or not.
- Once you started tracing and reload the page, the message should no longer appear.

### Issues:
- fixes #3363 

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [x] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable
- [x] Ready for review

daniel-wer commented on app/assets/javascripts/oxalis/api/api_latest.js:
```diff
@@ -790,7 +790,7 @@ class DataApi {
    *
    * @example
-   * api.user.setConfiguration("segmentationOpacity", 20);
+   * api.data.setConfiguration("segmentationOpacity", 20);
```
Some of these were messed up, probably because of copy paste errors :)

philippotto commented on app/assets/javascripts/oxalis/model/sagas/task_saga.js:
```diff
@@ -0,0 +1,101 @@
+  if (confirmed) {
+    for (const key of Object.keys(settingsObject)) {
+      if (key === "zoom") {
+        yield* put(setZoomStepAction(settingsObject[key]));
```
Do you know why zoom is not held by the user configuration? It feels a bit weird to have this special handling here (and further above).

daniel-wer replied:
We talked about this in person, but for documentation purposes, zoom is neither a dataset nor a user setting, but instead persisted on a per-tracing basis (much like the edit position).

philippotto commented on app/assets/javascripts/test/enzyme/snapshot.e2e.js:
```diff
@@ -18,7 +18,6 @@ import mockRequire from "mock-require";
 import test from "ava";
 
 // Those wrappers interfere with global.window and global.document otherwise
-mockRequire("libs/keyboardjs_wrapper", {});
```
Are you sure that this change is necessary? The e2e tests failed on circleci :thinking:

daniel-wer replied:
The `keyboardjs_wrapper` no longer exists, the e2e tests failed because the schema.json couldn't be found, which I fixed in the mean time :)

philippotto commented:
Ok, I just tested the feature and it worked fine except for a few bits:

- when editing an existing task type, checking the checkbox for recommended settings yielded in an empty settings field (I tried it for the default task type and was wondering what I should type in)
- when using the "finish and get next task", I noticed two things:
  - switching to the next task without changing task type or dataset, the modal still appeared (even though  you wrote "The message will not trigger twice if the "Finish and get next task" functionality is used").
  - in addition, in the above scenario the modal appeared twice in a row even if I did not change the accepted settings. I stepped through the code a bit and found that the keyboard delay (which should be set to 0) seemed to differ in the store? However, the settings should the correct value :thinking: Very strange. You should be able to reproduce it by creating two equal tasks and using finish-and-get-next-task (click accept and don't change settings).

> The message will not trigger twice if the "Finish and get next task" functionality is used and the user changes settings in the mean time.

Where is this modelled in the code? I couldn't find the logic for that. I'm only seeing that on each wk_ready event, you are comparing the recommended settings with the actual ones? Wouldn't it make sense to reuse the "did task type change?" logic? If you are not changing the task type, the modal should probably never appear?

daniel-wer commented:
@philippotto Thanks for the thorough testing!
- When editing an existing task type without recommended settings, the default settings are now shown when checking the checkbox. I've also added all setting possibilities to the right of the textarea.
- Using "Finish and Get Next Task" should no longer re-trigger the recommended settings modal.
- The `keyboardDelay` bug was triggered because mistakenly the datasetConfiguration contained a keyboardDelay value as well (probably some legacy property), although it belongs into the userConfiguration. We've already fixed this in the DB on master. You can fix it locally by running this script in pgadmin: `update webknossos.user_datasetconfigurations set configuration = configuration - 'keyboardDelay'`.

It would be really cool if you could do another round of testing :)

daniel-wer commented:
@rschwanhold In order to make this feature even more useful we want to persist the `lastTaskTypeId` for each user. `lastTaskTypeId` should be an optional field for each user (so it doesn't need to be included when creating or updating the user). In order to make it foolproof to update the `lastTaskTypeId`, it would be cool if you could add a PATCH route (or probably better make the `/api/users/<userId>` route accept PATCH requests), meaning we could send a request that only contains the userId and `lastTaskTypeId` (or whatever other user properties) and only those properties are updated :)

rschwanhold commented:
@daniel-wer I have added the `lastTaskTypeId` and changed the `update`-method of the User to support JSON-files with only a few values.
The route is still labeled as `PUT` instead of `PATCH` because I didn't want to update the corresponding frontend code, but feel free to do it yourself.

rschwanhold commented on app/models/user/UserService.scala:
```diff
@@ -84,7 +84,8 @@ class UserService @Inject()(conf: WkConf,
         isSuperUser = false,
-        isDeactivated = !isActive
+        isDeactivated = !isActive,
+        lastTaskTypeId = None
```
@fm3 why do I need to specify `lastTaskTypeId` here but not in `InitialDataController` ?

fm3 replied:
Do you get compiler errors if you don’t? I don’t see why it would be a problem to omit it.

fm3 replied:
taking that back, I got the error also for InitialDataController. I suppose this has something to do with the other named parameters

daniel-wer commented:
@rschwanhold or @fm3 Seems to work very well, except that when getting the user, the lastTaskTypeId is an object with an id property instead of just the id as a string: `lastTaskTypeId: {id: "5be99ac0930100d330dfef94"}`. Could you change this, so it is just the string?

rschwanhold commented:
@daniel-wer you said that `lastTaskTypeId` should be optional. Is there a use case to set the field to `NULL` (in the DB) after this field contained an actual taskTypeId? Because I think right now this is not supported. How would you intend to set the value to `NULL` when you call the `update`-method?

daniel-wer commented:
@rschwanhold I don't think we'll ever need to set the `lastTaskTypeId` back to null after it contained a valid value. With "optional" I meant that it does not need to be included when updating the user. So all good :)

Do you know why the schema comparison fails on CircleCI?

daniel-wer commented:
> - When editing an existing task type without recommended settings, the default settings are now shown when checking the checkbox. I've also added all setting possibilities to the right of the textarea.
> - Using "Finish and Get Next Task" should no longer re-trigger the recommended settings modal.
> - The keyboardDelay bug was triggered because mistakenly the datasetConfiguration contained a keyboardDelay value as well (probably some legacy property), although it belongs into the userConfiguration. We've already fixed this in the DB on master. You can fix it locally by running this script in pgadmin: update webknossos.user_datasetconfigurations set configuration = configuration - 'keyboardDelay'.

@philippotto In addition to these changes, the lastTaskTypeId of a user is now persisted as well, which allows to really only show the "Task-Type Description" and "Recommended Configuration" Modals if the user traces a different task type than before :)

It would be great if you could have a final look and retest the things that failed for you last time :)

daniel-wer commented:
@rschwanhold nevermind, I fixed the schema stuff, now scalafmt complains. Would be cool if you could fix that :)
